src/ui/textfield.rs

```diff
@@ -85,56 +143,191 @@ impl TextField {
+                        (selection_anchor_text_width.round(), (Self::height(ui) * 0.2).round()),
+                        (selection_text_width.round(), (Self::height(ui) * 0.6).round()),
+                    ),
+                    0.0,
+                    0.0,
+                );
+                canvas.draw_rrect(rrect, &paint);
+            });
+        }
+
         canvas.restore();
         ui.pop_group();
 
-        // process events
+        // Process events
         self.process_events(ui, input);
 
         ui.pop_group();
     }
 
+    // Get selection content
+    fn selection_text(&self) -> String {
+        if self.selection.len() == 0 {
+            return String::new()
+        }
+
+        self.text[self.selection.normalize()].iter().collect()
+    }
+
+    // Set text
+    fn set_text(&mut self, text: String) {
+        self.text = text.chars().collect();
+        self.update_utf8();
+
+        self.selection.move_cursors(self.text.len());
+    }
+
     /// Resets the text field's blink timer.
     fn reset_blink(&mut self, input: &Input) {
         self.blink_start = input.time_in_seconds();
     }
 
-    /// Appends a character to the end of the text.
+    /// Appends a character to the cursor position.
+    /// Or replaces selection if any.
     fn append(&mut self, ch: char) {
-        self.text.push(ch);
+        if self.selection.len() > 0 {
+            self.text.splice(self.selection.normalize(), vec![ch]);
+
+            self.selection.move_cursors(self.selection.start() + 1);
+        } else {
+            self.text.insert(self.selection.first, ch);
+            self.selection.move_cursors_right();
+        }
+
         self.update_utf8();
     }
 
-    /// Removes a character from the end of the text.
+    /// Removes a character at cursor position.
+    /// Or removes selection if any.
     fn backspace(&mut self) {
-        self.text.pop();
+        if self.selection.len() != 0 {
+            self.delete();
+        } else if self.selection.first > 0 {
+            self.selection.move_cursors_left();
+            self.text.remove(self.selection.first);
+        }
+
+        self.update_utf8();
+    }
+
+    /// Removes character after cursor position.
+    /// Or selection if any.
+    fn delete(&mut self) {
+        if self.selection.len() != 0 {
+            self.text.drain(self.selection.normalize());
+            self.selection.move_cursors(self.selection.start());
+        } else if self.selection.first != self.text.len() {
+            self.text.remove(self.selection.first);
+        }
+
         self.update_utf8();
     }
 
+    fn key_ctrl_down(&self, input: &Input) -> bool {
+        input.key_is_down(VirtualKeyCode::LControl) || input.key_is_down(VirtualKeyCode::RControl)
+    }
+
+    fn key_shift_down(&self, input: &Input) -> bool {
+        input.key_is_down(VirtualKeyCode::LShift) || input.key_is_down(VirtualKeyCode::RShift)
+    }
+
+    fn handle_word_skipping_and_selection(
```

Comment:
Generally NetCanv's codebase uses the word `process` for processing input and UI layout, rather than `handle`.